src/IR/IR.hs

```diff
@@ -186,6 +190,12 @@ collectLambda (Lambda a b _) =
   let (as, body) = collectLambda b in (a : as, body)
 collectLambda e = ([], e)
 
+-- | Create a lambda chain given a list of argument-type pairs and a body.
+makeLambdaChain :: TypeSystem t => [(Binder, t)] -> Expr t -> Expr t
+makeLambdaChain []   body = body
```

Comment:
You don't need this base case anymore; `foldr` handles that for you

Reply:
oops, forgot to remove that, ty!